src/components/icon/icon-registry.ts

```diff
@@ -0,0 +1,370 @@
+      SVGElement {
+    // Iterate backwards, so icon sets added later have precedence.
+    for (let i = iconSetConfigs.length - 1; i >= 0; i--) {
+      const config = iconSetConfigs[i];
+      if (config.svgElement) {
+        const foundIcon = this._extractSvgIconFromSet(config.svgElement, iconName, config);
+        if (foundIcon) {
+          return foundIcon;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Loads the content of the icon URL specified in the SvgIconConfig and creates an SVG element
+   * from it.
+   */
+  private _loadSvgIconFromConfig(config: SvgIconConfig): Observable<SVGElement> {
+    return this._fetchUrl(config.url)
+        .map(svgText => this._createSvgElementForSingleIcon(svgText, config));
+  }
+
+  /**
+   * Loads the content of the icon set URL specified in the SvgIconConfig and creates an SVG element
+   * from it.
+   */
+  private _loadSvgIconSetFromConfig(config: SvgIconConfig): Observable<SVGElement> {
+    return this._fetchUrl(config.url)
+        .map((svgText) => this._svgElementFromString(svgText));
+  }
+
+  /**
+   * Creates a DOM element from the given SVG string, and adds default attributes.
+   */
+  private _createSvgElementForSingleIcon(responseText: string, config: SvgIconConfig): SVGElement {
+    const svg = this._svgElementFromString(responseText);
+    this._setSvgAttributes(svg, config);
+    return svg;
+  }
+
+  /**
+   * Searches the cached element of the given SvgIconConfig for a nested icon element whose "id"
+   * tag matches the specified name. If found, copies the nested element to a new SVG element and
+   * returns it. Returns null if no matching element is found.
+   */
+  private _extractSvgIconFromSet(
+      iconSet: SVGElement, iconName: string, config: SvgIconConfig): SVGElement {
+    const iconNode = iconSet.querySelector('#' + iconName);
+    if (!iconNode) {
+      return null;
+    }
+    // If the icon node is itself an <svg> node, clone and return it directly. If not, set it as
+    // the content of a new <svg> node.
+    if (iconNode.tagName.toLowerCase() == 'svg') {
+      return this._setSvgAttributes(<SVGElement>iconNode.cloneNode(true), config);
+    }
+    // createElement('SVG') doesn't work as expected; the DOM ends up with
+    // the correct nodes, but the SVG content doesn't render. Instead we
+    // have to create an empty SVG node using innerHTML and append its content.
+    // http://stackoverflow.com/questions/23003278/svg-innerhtml-in-firefox-can-not-display
+    const svg = this._svgElementFromString('<svg></svg>');
+    // Clone the node so we don't remove it from the parent icon set element.
+    svg.appendChild(iconNode.cloneNode(true));
+    return this._setSvgAttributes(svg, config);
+  }
+
+  /**
+   * Creates a DOM element from the given SVG string.
+   */
+  private _svgElementFromString(str: string): SVGElement {
+    // TODO: Is there a better way than innerHTML? Renderer doesn't appear to have a method for
```

Comment:
I've played with this for a while, but it appears that using DOMParser to create an <svg> element has the same problem as document.createElement('svg'), where the DOM content looks correct but the icon doesn't render. (similar to the referenced http://stackoverflow.com/questions/23003278/svg-innerhtml-in-firefox-can-not-display)